pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -338,15 +342,117 @@ func (pl *VolumeRestrictions) EventsToRegister() []framework.ClusterEventWithHin
 		// Once running pods are deleted and volumes have been released, the unschedulable pod will be schedulable.
 		// Due to immutable fields `spec.volumes`, pod update events are ignored.
-		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}},
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}, QueueingHintFn: pl.isSchedulableAfterPodDeleted},
 		// A new Node may make a pod schedulable.
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterNodeAdded},
 		// Pods may fail to schedule because the PVC it uses has not yet been created.
 		// This PVC is required to exist to check its access modes.
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeClaimChange},
 	}
 }
 
+// isSchedulableAfterPodDeleted is invoked whenever a pod deleted,
+// It checks whether the deleted pod will conflict with volumes of other pods on the same node
+func (pl *VolumeRestrictions) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	oldPod, _, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterPodDeleted", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	nodeInfo := framework.NewNodeInfo(oldPod)
+	if !satisfyVolumeConflicts(pod, nodeInfo) {
+		return framework.QueueSkip
+	}
+
+	return framework.QueueImmediately
+}
+
+// isSchedulableAfterNodeAdded is invoked whenever a node added, It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeRestrictions) isSchedulableAfterNodeAdded(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	_, newNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeAdded", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	logger.V(4).Info("node was added", "pod", klog.KObj(pod), "node", klog.KObj(newNode))
+	return framework.QueueImmediately
+}
+
+// isSchedulableAfterPersistentVolumeClaimChange is invoked whenever a PersistentVolumeClaim changed, It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeRestrictions) isSchedulableAfterPersistentVolumeClaimChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	if newObj == nil {
+		// Deletes don't make a pod schedulable.
```

Comment:
I don't think there will be a delete event, we didn't register a pvc delete event.
We need to check the `add` event and `update` event here. For add event, we need to check if this pvc is `ReadWriteOncePod` and used by other pod. For update event, we need to check if it changes the access mode or satisfies VolumeConflict

Reply:
> I don't think there will be a delete event, we didn't register a pvc delete event. We need to check the `add` event and `update` event here. For add event, we need to check if this pvc is `ReadWriteOncePod` and used by other pod. For update event, we need to check if it changes the access mode or satisfies VolumeConflict

Yeah, it's a mistake.

Reply:
> we need to check if this pvc is `ReadWriteOncePod` and used by other pod

I think for both `update` and add `event`, we should do this check



> we need to check if it changes the access mode or satisfies VolumeConflict

we shouldn't check for `VolumeConflict` because it's not relate with the pod's pvc